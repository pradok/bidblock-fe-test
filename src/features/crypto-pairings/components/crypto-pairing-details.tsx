import * as React from 'react';
import {connect} from 'react-redux';
import {Card, Spin} from 'antd';
import {toUpper} from 'lodash';
// @ts-ignore
import * as Icon from 'react-cryptocoins';

import {tickerRawToDetail} from '../../../utils/transformer';
import {RootState} from '../../../store';

import {TickerDetails} from '../models';
import {fetchCryptoTicker} from '../api';

import {splitCryptoPairingSymbol, numToLocale} from '../../../utils/transformer';

interface StateProps {
  tickerDetails: TickerDetails;
  tickerFetch: boolean;
}

interface State {
  tickerDetails: TickerDetails;
  isLoading: boolean;
}

interface OwnProps {
  tickerIndex: string;
  tickerSymbol: string;
}

type Props = StateProps & OwnProps;

const tickerDetailsDefault = {
  ask: 0,
  askSize: 0,
  bid: 0,
  bidSize: 0,
  dailyChange: 0,
  dailyChangePerc: 0,
  high: 0,
  lastPrice: 0,
  low: 0,
  volume: 0,
};

class CryptoPairingDetails extends React.Component<Props, State> {

  public static defaultProps: Partial<Props> = {
    tickerDetails: tickerDetailsDefault,
    tickerFetch: false,
  };
  
  constructor(props: Props){
    super(props);
    this.state = { isLoading: false, tickerDetails: props.tickerDetails };
  }

  public componentDidMount() {
    const {tickerFetch, tickerSymbol} = this.props;
    if (tickerFetch && tickerSymbol) {
      this.setState({isLoading: true});
      fetchCryptoTicker(tickerSymbol).then(data => this.setState(
        {tickerDetails: tickerRawToDetail(data), isLoading: false}))
        .catch(reason => console.log('reason:', reason.message));
    }
  }

  public renderDetails() {
    const {tickerDetails} = this.state;
    const cryptos = splitCryptoPairingSymbol(this.props.tickerSymbol);
    const CryptoIcon = Icon[cryptos[0]];
    return (
      <Spin spinning={this.state.isLoading} size="large">
        <Card title="Ticker Details">
          <h2>{CryptoIcon && <CryptoIcon size={48} />} {toUpper(`${cryptos[0]}/${cryptos[1]}`)}</h2>
          <h4>Bid</h4>
          <p>{numToLocale(tickerDetails.bid)}</p>
          <h4>Bid size</h4>
          <p>{numToLocale(tickerDetails.bidSize)}</p>
          <h4>Ask</h4>
          <p>{numToLocale(tickerDetails.ask)}</p>
          <h4>Ask size</h4>
          <p>{numToLocale(tickerDetails.askSize)}</p>
          <h4>Daily change</h4>
          <p>{numToLocale(tickerDetails.dailyChange)}</p>
          <h4>Last price</h4>
          <p>{numToLocale(tickerDetails.lastPrice)}</p>
          <h4>Volume</h4>
          <p>{numToLocale(tickerDetails.volume)}</p>
        </Card>
      </Spin>
    )
  }

  public render() {
    return (
      <React.Fragment>
        {this.renderDetails()}
      </React.Fragment>
    )
  }
}

const mapStateToProps = (state: RootState, props: OwnProps) => {
  const {tickerCollection} = state.cryptoPairings;
  return {
    tickerDetails: tickerCollection[props.tickerIndex],
    tickerFetch: !tickerCollection[props.tickerIndex]
  }
};

export const CryptoTickerDetails = CryptoPairingDetails ;

export default connect<StateProps, {}, {}>(
  mapStateToProps,
  {}
)(CryptoPairingDetails);

