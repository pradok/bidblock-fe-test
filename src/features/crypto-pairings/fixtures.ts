import {CryptoPairingsCollection, TickerDetails} from "./models";

const pairings: CryptoPairingsCollection = [
  [
    "tBTCUSD", 6239.9, 34.92908368, 6240, 35.7623858, -30.1, -0.0048, 6239.9, 18771.42426519, 6339.9, 6125
  ],
  [
    "tLTCUSD", 75.999, 1099.6280602, 76, 486.87405918, -1.648, -0.0212, 76, 110391.68659301, 78.66, 75.049
  ]
];

const tickerDetails: TickerDetails = {
  ask: 0.0002709,
  askSize: 43891.54842947,
  bid: 0.00021887,
  bidSize: 1553050.60274823,
  dailyChange: -0.00000617,
  dailyChangePerc: -0.0267,
  high: 0.000232,
  lastPrice: 0.000225,
  low: 0.000225,
  volume: 48787.142,
};

export {tickerDetails, pairings};