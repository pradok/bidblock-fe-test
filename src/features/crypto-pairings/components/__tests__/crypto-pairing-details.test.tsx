import * as React from 'react';
import {shallow} from 'enzyme';

import {tickerDetails} from '../../fixtures';

import {CryptoTickerDetails} from '../crypto-pairing-details';

it('<CryptoTickerDetails> renders with Ticker Details', () => {
  const component = shallow(
    <CryptoTickerDetails
      tickerDetails={tickerDetails}
      tickerFetch={false}
      tickerIndex="1"
      tickerSymbol="tBTCUSD"
    />
  );
  expect(component).toMatchSnapshot();
});
