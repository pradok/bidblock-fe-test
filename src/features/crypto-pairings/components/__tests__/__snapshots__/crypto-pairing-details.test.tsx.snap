// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CryptoTickerDetails> renders with Ticker Details 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CryptoPairingDetails
    tickerDetails={
      Object {
        "ask": 0.0002709,
        "askSize": 43891.54842947,
        "bid": 0.00021887,
        "bidSize": 1553050.60274823,
        "dailyChange": -0.00000617,
        "dailyChangePerc": -0.0267,
        "high": 0.000232,
        "lastPrice": 0.000225,
        "low": 0.000225,
        "volume": 48787.142,
      }
    }
    tickerFetch={false}
    tickerIndex="1"
    tickerSymbol="tBTCUSD"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Spin
        prefixCls="ant-spin"
        size="large"
        spinning={false}
        wrapperClassName=""
      >
        <Card
          title="Ticker Details"
        >
          <h2>
            <Btc
              color="currentColor"
              size={48}
            />
             
            BTC/USD
          </h2>
          <h4>
            Bid
          </h4>
          <p>
            0.00021887
          </p>
          <h4>
            Bid size
          </h4>
          <p>
            1553050.60274823
          </p>
          <h4>
            Ask
          </h4>
          <p>
            0.0002709
          </p>
          <h4>
            Ask size
          </h4>
          <p>
            43891.54842947
          </p>
          <h4>
            Daily change
          </h4>
          <p>
            -0.00000617
          </p>
          <h4>
            Last price
          </h4>
          <p>
            0.000225
          </p>
          <h4>
            Volume
          </h4>
          <p>
            48787.142
          </p>
        </Card>
      </Spin>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Card
          title="Ticker Details"
        >
          <h2>
            <Btc
              color="currentColor"
              size={48}
            />
             
            BTC/USD
          </h2>
          <h4>
            Bid
          </h4>
          <p>
            0.00021887
          </p>
          <h4>
            Bid size
          </h4>
          <p>
            1553050.60274823
          </p>
          <h4>
            Ask
          </h4>
          <p>
            0.0002709
          </p>
          <h4>
            Ask size
          </h4>
          <p>
            43891.54842947
          </p>
          <h4>
            Daily change
          </h4>
          <p>
            -0.00000617
          </p>
          <h4>
            Last price
          </h4>
          <p>
            0.000225
          </p>
          <h4>
            Volume
          </h4>
          <p>
            48787.142
          </p>
        </Card>,
        "prefixCls": "ant-spin",
        "size": "large",
        "spinning": false,
        "wrapperClassName": "",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <h2>
              <Btc
                color="currentColor"
                size={48}
              />
               
              BTC/USD
            </h2>,
            <h4>
              Bid
            </h4>,
            <p>
              0.00021887
            </p>,
            <h4>
              Bid size
            </h4>,
            <p>
              1553050.60274823
            </p>,
            <h4>
              Ask
            </h4>,
            <p>
              0.0002709
            </p>,
            <h4>
              Ask size
            </h4>,
            <p>
              43891.54842947
            </p>,
            <h4>
              Daily change
            </h4>,
            <p>
              -0.00000617
            </p>,
            <h4>
              Last price
            </h4>,
            <p>
              0.000225
            </p>,
            <h4>
              Volume
            </h4>,
            <p>
              48787.142
            </p>,
          ],
          "title": "Ticker Details",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Btc
                  color="currentColor"
                  size={48}
                />,
                " ",
                "BTC/USD",
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "color": "currentColor",
                  "size": 48,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              " ",
              "BTC/USD",
            ],
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Bid",
            },
            "ref": null,
            "rendered": "Bid",
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 0.00021887,
            },
            "ref": null,
            "rendered": 0.00021887,
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Bid size",
            },
            "ref": null,
            "rendered": "Bid size",
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 1553050.60274823,
            },
            "ref": null,
            "rendered": 1553050.60274823,
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Ask",
            },
            "ref": null,
            "rendered": "Ask",
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 0.0002709,
            },
            "ref": null,
            "rendered": 0.0002709,
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Ask size",
            },
            "ref": null,
            "rendered": "Ask size",
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 43891.54842947,
            },
            "ref": null,
            "rendered": 43891.54842947,
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Daily change",
            },
            "ref": null,
            "rendered": "Daily change",
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": -0.00000617,
            },
            "ref": null,
            "rendered": -0.00000617,
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Last price",
            },
            "ref": null,
            "rendered": "Last price",
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 0.000225,
            },
            "ref": null,
            "rendered": 0.000225,
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Volume",
            },
            "ref": null,
            "rendered": "Volume",
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 48787.142,
            },
            "ref": null,
            "rendered": 48787.142,
            "type": "p",
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Spin
          prefixCls="ant-spin"
          size="large"
          spinning={false}
          wrapperClassName=""
        >
          <Card
            title="Ticker Details"
          >
            <h2>
              <Btc
                color="currentColor"
                size={48}
              />
               
              BTC/USD
            </h2>
            <h4>
              Bid
            </h4>
            <p>
              0.00021887
            </p>
            <h4>
              Bid size
            </h4>
            <p>
              1553050.60274823
            </p>
            <h4>
              Ask
            </h4>
            <p>
              0.0002709
            </p>
            <h4>
              Ask size
            </h4>
            <p>
              43891.54842947
            </p>
            <h4>
              Daily change
            </h4>
            <p>
              -0.00000617
            </p>
            <h4>
              Last price
            </h4>
            <p>
              0.000225
            </p>
            <h4>
              Volume
            </h4>
            <p>
              48787.142
            </p>
          </Card>
        </Spin>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Card
            title="Ticker Details"
          >
            <h2>
              <Btc
                color="currentColor"
                size={48}
              />
               
              BTC/USD
            </h2>
            <h4>
              Bid
            </h4>
            <p>
              0.00021887
            </p>
            <h4>
              Bid size
            </h4>
            <p>
              1553050.60274823
            </p>
            <h4>
              Ask
            </h4>
            <p>
              0.0002709
            </p>
            <h4>
              Ask size
            </h4>
            <p>
              43891.54842947
            </p>
            <h4>
              Daily change
            </h4>
            <p>
              -0.00000617
            </p>
            <h4>
              Last price
            </h4>
            <p>
              0.000225
            </p>
            <h4>
              Volume
            </h4>
            <p>
              48787.142
            </p>
          </Card>,
          "prefixCls": "ant-spin",
          "size": "large",
          "spinning": false,
          "wrapperClassName": "",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <h2>
                <Btc
                  color="currentColor"
                  size={48}
                />
                 
                BTC/USD
              </h2>,
              <h4>
                Bid
              </h4>,
              <p>
                0.00021887
              </p>,
              <h4>
                Bid size
              </h4>,
              <p>
                1553050.60274823
              </p>,
              <h4>
                Ask
              </h4>,
              <p>
                0.0002709
              </p>,
              <h4>
                Ask size
              </h4>,
              <p>
                43891.54842947
              </p>,
              <h4>
                Daily change
              </h4>,
              <p>
                -0.00000617
              </p>,
              <h4>
                Last price
              </h4>,
              <p>
                0.000225
              </p>,
              <h4>
                Volume
              </h4>,
              <p>
                48787.142
              </p>,
            ],
            "title": "Ticker Details",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Btc
                    color="currentColor"
                    size={48}
                  />,
                  " ",
                  "BTC/USD",
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "color": "currentColor",
                    "size": 48,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                " ",
                "BTC/USD",
              ],
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Bid",
              },
              "ref": null,
              "rendered": "Bid",
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 0.00021887,
              },
              "ref": null,
              "rendered": 0.00021887,
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Bid size",
              },
              "ref": null,
              "rendered": "Bid size",
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 1553050.60274823,
              },
              "ref": null,
              "rendered": 1553050.60274823,
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Ask",
              },
              "ref": null,
              "rendered": "Ask",
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 0.0002709,
              },
              "ref": null,
              "rendered": 0.0002709,
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Ask size",
              },
              "ref": null,
              "rendered": "Ask size",
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 43891.54842947,
              },
              "ref": null,
              "rendered": 43891.54842947,
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Daily change",
              },
              "ref": null,
              "rendered": "Daily change",
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": -0.00000617,
              },
              "ref": null,
              "rendered": -0.00000617,
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Last price",
              },
              "ref": null,
              "rendered": "Last price",
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 0.000225,
              },
              "ref": null,
              "rendered": 0.000225,
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Volume",
              },
              "ref": null,
              "rendered": "Volume",
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 48787.142,
              },
              "ref": null,
              "rendered": 48787.142,
              "type": "p",
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
